### PR TITLE
Add recipe for libignition-sensors4

### PR DESCRIPTION
**Pull request by traversaro**
Related to https://github.com/conda-forge/staged-recipes/issues/13551 . This adds a new missing library out of the ignition robotics libraries (see https://ignitionrobotics.org/).

The build on macOS is failing due to the need to 10.13 SDK, as done in https://github.com/conda-forge/staged-recipes/pull/13677 we can merge the recipe and then re-render in the newly created feedstock.

~I tried to put the recipe in the `libignition-sensors` directory even if it creates a `libignition-sensors4` package as given that we use a single feedstock to generate all the (different) packages named `libignition-sensors<n>` we would like it to be called `libignition-sensors-feedstock` instead of `libignition-sensors4-feedstock` (as in the future it will host also `libignition-sensors5`, `libignition-sensors5`, ...) but to be honest I am not sure it will work. However, this is just a nice to have, nothing really important.~ Actually we modified the feedstock to specify explicitly its name, see https://github.com/conda-forge/staged-recipes/pull/14313#discussion_r598149749 .

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by traversaro** [comment]
@wolfv @Tobias-Fischer @j-rivero @JShep1 I listed yourself as additional maintainers, please confirm if this is ok for you, or if you prefer not to be a mantainer of this package.

**Comment by wolfv** [comment]
I'm fine with that!

**Comment by traversaro** [comment]
macOS fails with:
~~~
In file included from /Users/runner/miniforge3/conda-bld/libignition-sensors4_1616242492432/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/include/sdformat-10.3/sdf/../sdf/Element.hh:28:
/Users/runner/miniforge3/conda-bld/libignition-sensors4_1616242492432/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/include/sdformat-10.3/sdf/../sdf/Param.hh:350:37: error: 'get<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, bool, char, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, int, unsigned long long, unsigned int, double, float, sdf::v10::Time, ignition::math::v6::Angle, ignition::math::v6::Color, ignition::math::v6::Vector2<int>, ignition::math::v6::Vector2<double>, ignition::math::v6::Vector3<double>, ignition::math::v6::Quaternion<double>, ignition::math::v6::Pose3<double>>' is unavailable: introduced in macOS 10.13
        std::string strValue = std::get<std::string>(this->dataPtr->value);
                                    ^
/Users/runner/miniforge3/conda-bld/libignition-sensors4_1616242492432/_build_env/bin/../include/c++/v1/variant:1435:16: note: 'get<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, bool, char, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, int, unsigned long long, unsigned int, double, float, sdf::v10::Time, ignition::math::v6::Angle, ignition::math::v6::Color, ignition::math::v6::Vector2<int>, ignition::math::v6::Vector2<double>, ignition::math::v6::Vector3<double>, ignition::math::v6::Quaternion<double>, ignition::math::v6::Pose3<double>>' has been explicitly marked unavailable here
constexpr _Tp& get(variant<_Types...>& __v) {
~~~

We probably need to target 10.13 due to the use of recent C++ features.

**Comment by traversaro** [comment]
@conda-forge/help-c-cpp The PR is now ready for review. See the fist comment for an overview, and for the explanation of why the macOS build is failing and it make sense to merge the PR regardless.

**Comment by traversaro** [comment]
> 
> 
> @wolfv @Tobias-Fischer @j-rivero @JShep1 I listed yourself as additional maintainers, please confirm if this is ok for you, or if you prefer not to be a mantainer of this package.

For the time being I removed your names, once the PR is merged we can add them back if you want, thanks!

**Comment by Tobias-Fischer** [comment]
I'm happy, too. One remark, last time we merged an ignition feedstock we were wondering if there's a way to avoid having the version number in the feedstock name. Should we try here?

File: recipes/libignition-sensors/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    - if exist $PREFIX$\\Library\\bin\\ignition-{{ component_name }}{{ major_version }}.dll (exit 0) else (exit 1)  # [win]
+    - if exist %PREFIX%\\Library\\lib\\cmake\\ignition-{{ component_name }}{{ major_version }}\\ignition-{{ component_name }}{{ major_version }}-config.cmake (exit 0) else (exit 1)  # [win]
+
+about:
+  home: https://github.com/ignitionrobotics/ign-{{ component_name }}
+  license: Apache-2.0
+  license_file: LICENSE
+  summary: Ignition Robotics library that provide numerous sensor models designed to generate realistic data from simulation environments. 
+
+extra:
```
Comment (xhochy) [inline recipes/libignition-sensors/meta.yaml:57]: This is how you get the number out of the feedstock name.
```suggestion
extra:
  feedstock-name: libignition-sensors
```

**Comment by j-rivero** [comment]
> For the time being I removed your names, once the PR is merged we can add them back if you want, thanks!

fine for me too. Thanks Silvio

File: recipes/libignition-sensors/meta.yaml (hunk 1 above)
Comment (traversaro) [reply recipes/libignition-sensors/meta.yaml:57]: Thanks!

**Comment by traversaro** [comment]
> 
> 
> I'm happy, too. One remark, last time we merged an ignition feedstock we were wondering if there's a way to avoid having the version number in the feedstock name. Should we try here?

@xhochy provided the solution, it is just sufficient to specify the `extra.feedstock_name` parameter (see https://github.com/conda-forge/staged-recipes/pull/14313#issuecomment-803450148).

**Comment by traversaro** [comment]
Thanks @Tobias-Fischer @j-rivero , I added you as maintainers.

**Comment by wolfv** [comment]
I think the tests on windows should use if not exists ... exit 1

Happy to merge right after!

**Comment by traversaro** [comment]
> 
> 
> I think the tests on windows should use if not exists ... exit 1
> 
> Happy to merge right after!

Old habits die hard. : ) Done! I also tried to make the recipe quite parametric, so for future ignition libraries it should be sufficient to just copy the meta.yaml and change dependencies and initial metadata.

File: recipes/libignition-sensors4/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    - __osx >={{ MACOSX_DEPLOYMENT_TARGET|default("10.9") }}  # [osx and x86_64]
+
+test:
+  commands:
+    - test -f ${PREFIX}/include/ignition/{{ component_name }}{{ major_version }}/ignition/{{ component_name }}.hh  # [not win]
+    - test -f ${PREFIX}/lib/libignition-{{ component_name }}{{ major_version }}.so  # [linux]
+    - test -f ${PREFIX}/lib/libignition-{{ component_name }}{{ major_version }}.dylib  # [osx]
+    - test -f ${PREFIX}/lib/cmake/ignition-{{ component_name }}{{ major_version }}/ignition-{{ component_name }}{{ major_version }}-config.cmake  # [not win]
+    - if not exist %PREFIX%\\Library\\include\\ignition\\{{ component_name }}{{ major_version }}\\ignition\\{{ component_name }}.hh exit 1  # [win]
+    - if not exist $PREFIX$\\Library\\lib\\ignition-{{ component_name }}{{ major_version }}.lib (exit 0) exit 1  # [win]
```
Comment (wolfv) [inline recipes/libignition-sensors4/meta.yaml:47]: I think here is still a typo?
Instead of (exit 0) exit 1, it should just be (exit 1), right?

Comment (traversaro) [reply recipes/libignition-sensors4/meta.yaml]: Yes, fixed now, sorry for the noise.

**Comment by traversaro** [comment]
After the changes the Windows existence checks are failing, not sure why.

**Comment by wolfv** [comment]
yeah the build log is really inconclusive :/ . Let me restart the windows build

**Comment by traversaro** [comment]
> 
> 
> yeah the build log is really inconclusive :/ . Let me restart the windows build

There is the same failure in https://github.com/conda-forge/staged-recipes/pull/14323 that was copy pasted from this, so i guess there is some typo in the check, looking at it again.

**Comment by traversaro** [comment]
> 
> 
> > yeah the build log is really inconclusive :/ . Let me restart the windows build
> 
> There is the same failure in #14323 that was copy pasted from this, so i guess there is some typo in the check, looking at it again.

Found it, it was `$PREFIX$` instead of the correct `%PREFIX%`.

**Pull request merged**